### PR TITLE
Local storage && Cache

### PR DESCRIPTION
**Pull request by hsingh6764**
Local storage && Cache

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,13 +4,17 @@
 
 Installation:
 
-    brew install vips
+    brew install vips (For mac: brew install homebrew/science/vips --with-webp --with-graphicsmagick)
```
Comment (wlmoore-webs) [inline README.md]: this line is confusing - homebrew is Mac-only.  Were the original instructions just wrong?

File: config/development.json [hunk 2]
```diff
@@ -1,9 +1,12 @@
 {
-	"PROJECT_NAME": "node-url-imaging",
-	"HOSTS": ["127.0.0.1"],
-	"URL": "http://127.0.0.1:3000",
-	"PORT": 3000,
-	"DEBUG_PORT": 5858,
+	"projectName": "node-url-imaging",
+	"port": 3300,
+	"debug_port": 5757,
```
Comment (wlmoore-webs) [inline config/development.json]: why mix camel-case (```projectName```) and underscore-delimited (```debug_port```) config parameters?

**Comment by HackAttack** [comment]
Please allow more than 13 minutes for code review before merging pull requests.

File: README.md (hunk 1 above)
Comment (hsingh6764) [inline README.md]: there is no formula vips
so in-order to get it we point to different library. 

https://github.com/lovell/sharp

File: config/development.json (hunk 2 above)
Comment (hsingh6764) [inline config/development.json]: going to fix this soon. (In next pull request)

**Comment by hsingh6764** [comment]
@HackAttack u can go head and comment more, I will make the change in the next pull request.

**Comment by hsingh6764** [comment]
@wlmoore-webs @richardadjogah I was wondering what if we run this in dev instead of other image processor.

File: src/imageManagement.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+var localStorage = require('./storage/localStorage');
+var logger = require('../config/log');
+
+var Storage = {
+	"local": localStorage
+};
+
+module.exports = {
+	getProcessedImage: function (fileName, callback) {
+		Storage[config.get('storage').type].getImage(fileName, function (err, data) {
```
Comment (wlmoore-webs) [inline src/imageManagement.js]: we should probably just read this as a static value during the require block, i.e.
```
var storageType = require('../config').get('storage').type;
```
Then we can just change the Storage declaration to be redefined to use the correct storage type at declaration time and we can just do ```Storage.getImage(fileName, function ...```

File: src/imageManagement.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	writeImageFile: function (fileName, buffer, callback) {
+		Storage[config.get('storage').type].writeImageFile(fileName, buffer, callback);
+	},
+
+	getFileInfo: function (fileName, callback) {
+		Storage[config.get('storage').type].getFileInfo(fileName, callback);
+	}
+};
+
+var getFormatedUri = function (url) {
```
Comment (wlmoore-webs) [inline src/imageManagement.js]: misspelled (```getFormattedUri```)

File: README.md (hunk 1 above)
Comment (wlmoore-webs) [inline README.md]: so just change the text, leaving "brew install vips" as the commnad is wrong.

File: src/processImage.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	async.parallel([
+		function (callback) {
+			ImageManagement.getProcessedImage(fileName, function (data) {
+				callback(null, data);
+			});
+		}, function (callback) {
+			ImageManagement.getFileInfo(fileName, callback);
+		}
+	], function (err, results) {
+		var refreashProcessedImage = false;
```
Comment (wlmoore-webs) [inline src/processImage.js]: another mispelling ```refreshProcessedImage```

**Comment by richardadjogah** [comment]
@hsingh6764 The plan for rolling this out is to eventually put the new imageprocessor into the load balancer . In theory, it should work identically to the original app apart from the changes mentioned in the readme.

File: src/processImage.js (hunk 5 above)
Comment (hsingh6764) [inline src/processImage.js]: this happens a lot when i am working at night with little light. :)

File: src/processImage.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+				if (maxCacheTimeInMin < dif) {
+					refreashProcessedImage = true;
+				}
+			} else {
+				refreashProcessedImage = true;
+			}
+		}
+
+		if (!refreashProcessedImage) {
+			logger.info("Serving cached filed: " + fileName);
```
Comment (wlmoore-webs) [inline src/processImage.js]: I think you need more sleep (```filed```)

File: src/processImage.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+			res.set('Content-Type', 'image');
+			res.send(results[0]);
+		} else {
+			async.waterfall([
+				function (next) {
+					logger.info('downloading image');
```
Comment (wlmoore-webs) [inline src/processImage.js]: shouldn't we put the URL or something else in here for tracking?

File: src/processImage.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+				}, function (body, next) {
+					sharp(body).metadata(function (err, metadata) {
+						next(null, body, metadata);
+					});
+				}, function (body, metadata, next) {
+					var sharpobject = sharp(body).quality(QUALITY);
+					var did_op = true;
+
+					// Iterate over list of commands
+					while (did_op) {
```
Comment (wlmoore-webs) [inline src/processImage.js]: I know this is copy-paste, but let's take a closer look at this.  I don't think this pattern is a particularly good solution for our problem - we really should be parsing the command line and taking operations, not just looping over all possible commands and seeing which ones work.  That could lead to all kinds of weird edge cases.

File: src/processImage.js [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+						}
+					}
+					sharpobject.toBuffer(next);
+				}, function (buffer, info, next) {
+					ImageManagement.writeImageFile(fileName, buffer, next);
+				}
+			], function (err, data) {
+
+				if (err) {
+					logger.error(err);
```
Comment (wlmoore-webs) [inline src/processImage.js]: probably should send a response for an error?

File: src/processImage.js (hunk 8 above)
Comment (richardadjogah) [inline src/processImage.js]: I agree that in practice this probably isn't the best solution, but I opted to replicate the logic from the original Django app to try to ensure functionality is as close to the original as possible.

File: src/processImage.js (hunk 9 above)
Comment (hsingh6764) [inline src/processImage.js]: i think we want to server broken img incase of error

File: src/processImage.js (hunk 8 above)
Comment (hsingh6764) [inline src/processImage.js]: when u get time @richardadjogah try to clean this up and I will also look at how we can improve this.

**Pull request merged**